crates/uv-client/src/base_client.rs

```diff
@@ -1139,8 +1139,66 @@ fn retryable_on_request_failure(err: &(dyn Error + 'static)) -> Option<Retryable
     None
 }
 
-pub fn is_transient_network_error(err: &(dyn Error + 'static)) -> bool {
-    retryable_on_request_failure(err) == Some(Retryable::Transient)
+/// Decide whether a requests should be retried, tracking previous attempts and backoff.
+pub struct RetryState {
+    retry_policy: ExponentialBackoff,
+    start_time: SystemTime,
+    total_retries: u32,
+    url: DisplaySafeUrl,
+}
+
+impl RetryState {
+    pub fn new(retry_policy: ExponentialBackoff, url: impl Into<DisplaySafeUrl>) -> Self {
+        Self {
+            retry_policy,
+            start_time: SystemTime::now(),
```

Comment:
Maybe it would be good to add a comment mentioning that new starts a timer? Otherwise someone might initialise this too early by accident.

Reply:
Good idea, renamed to `start`